### PR TITLE
Fix esm2 token classification metric and loss, add flip benchmark

### PR DESCRIPTION
**Pull request by yzhang123**
### Description

Fix esm2 token classification loss calculation and accuracy calculation, verified on flip benchmark 
add flip benchmark

### Type of changes

- [x]  Bug fix (non-breaking change which fixes an issue)
- [x]  New feature (non-breaking change which adds functionality)
- [ ]  Refactor
- [ ]  Documentation update
- [ ]  Other (please describe):

### CI Pipeline Configuration
Configure CI behavior by applying the relevant labels:

- [SKIP_CI](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#skip_ci) - Skip all continuous integration tests
- [INCLUDE_NOTEBOOKS_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_notebooks_tests) - Execute notebook validation tests in pytest
- [INCLUDE_SLOW_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_slow_tests) - Execute tests labelled as slow in pytest for extensive testing

> [!NOTE]
> By default, the notebooks validation tests are skipped unless explicitly enabled.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

* If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
* If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Usage

```python
TODO: Add code snippet
```

### Pre-submit Checklist

 - [x] I have tested these changes locally
 - [ ] I have updated the documentation accordingly
 - [x] I have added/updated tests as needed
 - [ ] All existing tests pass successfully

### CODE REVIEW

**Comment by trvachov** [comment]
@sichu2023 @polinabinder1 could you review please?

**Comment by yzhang123** [comment]
/ok to test [0ae16a6](https://github.com/NVIDIA/bionemo-framework/pull/946/commits/0ae16a602c0397bdf4e2bc6e420444b7564092c2)

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 1]
```diff
@@ -123,6 +160,46 @@ def __getitem__(self, index: int) -> BertSample:
...
+            tokens_no_special = torch.tensor(tokens_no_special)
+        tokens_no_special = tokens_no_special.flatten()
+
+        # Tokenize with special tokens
+        tokens_with_special = self.tokenizer.encode(sequence, add_special_tokens=True, return_tensors="pt")
+        if isinstance(tokens_with_special, list):
+            tokens_with_special = torch.tensor(tokens_with_special)
+        tokens_with_special = tokens_with_special.flatten()
+
+        # Find positions where special tokens were inserted
```
Comment (farhadrgh) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:185]: Is there an instance where special tokens are not at either end of the sequence?

**Review by farhadrgh — APPROVED** [review]
LGTM

**Comment by farhadrgh** [comment]
> /ok to test [0ae16a6](https://github.com/NVIDIA/bionemo-framework/commit/0ae16a602c0397bdf4e2bc6e420444b7564092c2)

@yzhang123 you need the whole commit hash after ok to test:

/ok to test 0ae16a602c0397bdf4e2bc6e420444b7564092c2

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 1 above)
Comment (yzhang123) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:185]: no, but just to be safe

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 2]
```diff
@@ -123,6 +160,46 @@ def __getitem__(self, index: int) -> BertSample:
...
+        special_positions = []
+
+        # Compare token by token to find insertions
+        original_idx = 0
+        for tokenized_idx, token_id in enumerate(tokens_with_special):
+            if token_id.item() in self.tokenizer.all_special_ids:
+                # This is a special token that was inserted
+                special_positions.append(tokenized_idx)
+            else:
```
Comment (dorotat-nv) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py]: shouldnt it be part of a unit test of this method rather than a sanity check in the tokenizer left for debuging?

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 3]
```diff
@@ -123,6 +160,46 @@ def __getitem__(self, index: int) -> BertSample:
...
+        if isinstance(tokens_with_special, list):
+            tokens_with_special = torch.tensor(tokens_with_special)
+        tokens_with_special = tokens_with_special.flatten()
+
+        # Find positions where special tokens were inserted
+        special_positions = []
+
+        # Compare token by token to find insertions
+        original_idx = 0
+        for tokenized_idx, token_id in enumerate(tokens_with_special):
```
Comment (dorotat-nv) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py]: this loop would be good for small tensors, but for larger ones, 
 a vectorised version should be much faster ie
 
```
 mask = torch.isin(tokens_with_special, special_ids_tensor)
special_positions = mask.nonzero(as_tuple=False).squeeze(1).tolist()
```

Comment (dorotat-nv) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py]: and the sanity check can be done

```
# Create mask for non-special tokens
non_special_mask = ~mask  # inverse of special mask

# Get positions of non-special tokens
non_special_positions = non_special_mask.nonzero(as_tuple=False).squeeze(1)

# Extract tokens from tokens_with_special which are not special
tokens_with_special_no_special = tokens_with_special[non_special_positions]
```

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 4]
```diff
@@ -123,6 +160,46 @@ def __getitem__(self, index: int) -> BertSample:
             "is_random": torch.zeros_like(tokenized_sequence, dtype=torch.int64),
         }
 
+    def _tokenize_with_special_tracking(self, sequence: str) -> tuple[Tensor, list[int]]:
```
Comment (dorotat-nv) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:163]: I would suggest to add a unit test for this

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 2 above)
Comment (yzhang123) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py]: done, thanks !

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 3 above)
Comment (yzhang123) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py]: done thanks!

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 4 above)
Comment (yzhang123) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:163]: done, thanks!

**Review by sichu2023 — APPROVED** [review]
Focused on reviewing metric logging and it looks good to me.

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 5]
```diff
@@ -108,12 +119,38 @@ def __len__(self) -> int:
...
+
+        # Handle labels_mask if it exists
+        if self.labels_mask is not None:
+            # Get the original labels_mask for this sequence
+            labels_mask = self.labels_mask.iloc[index]
+            labels_mask = list(labels_mask)
+
+            # Insert 0 tokens at special token positions to pad to tokenized_sequence length
+            idx = 0
+            while idx < len(special_positions):
```
Comment (dorotat-nv) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:134]: I am not sure what this code does, ie
having 

```
labels_mask = ['1', '1', '1']
special_positions = [1, 2]
```
A. 
should it output  ['1', '0', '0', '1', '1'] due to (current code- insert 0 for special tokens at indices of updated sequence)
```
labels_mask.insert(1, '0')  # ['1', '0', '1', '1']
labels_mask.insert(2, '0')  # Inserts at new index 2 → ['1', '0', '0', '1', '1'] 
```
or  B.
['1', '0',  '1',, "0" '1']  (insert 0 for special tokens at indices of old sequence )

```
labels_mask.insert(1, '0')  # ['1', '0', '1', '1']
labels_mask.insert(2, '0')  # Inserts at old index 2 → ['1', '0',  '1',, "0" '1'] 
```
Could you please externalise this method ie _update_labels_maks_with_special_tokens

If A is correct, we can simplify the "while statement" to what it currently does
```
for pos in special_positions:
    labels_mask.insert(pos, '0') 
```

if B is correct, then lets insert "0" from the last special tokens index to match the indexes between updated and old sequence

```
for pos in reversed(special_positions):
    labels_mask.insert(pos, '0')
```

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 6]
```diff
@@ -108,12 +119,38 @@ def __len__(self) -> int:
     def __getitem__(self, index: int) -> BertSample:
         """Obtains the BertSample at the given index."""
         sequence = self.sequences[index]
-        tokenized_sequence = self._tokenize(sequence)
+
+        # Tokenize sequence and track special token positions
+        tokenized_sequence, special_positions = self._tokenize_with_special_tracking(sequence)
+
+        # Handle labels_mask if it exists
+        if self.labels_mask is not None:
```
Comment (dorotat-nv) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:127]: I would suggest a change here 
```
  labels_mask = list(self.labels_mask.iloc[index])
```

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 7]
```diff
@@ -108,12 +119,38 @@ def __len__(self) -> int:
...
+            labels_mask = self.labels_mask.iloc[index]
+            labels_mask = list(labels_mask)
+
+            # Insert 0 tokens at special token positions to pad to tokenized_sequence length
+            idx = 0
+            while idx < len(special_positions):
+                labels_mask.insert(special_positions[idx], "0")
+                idx += 1
+            labels_mask = "".join(labels_mask)
+            assert len(labels_mask) == len(tokenized_sequence), (
```
Comment (dorotat-nv) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:138]: Assert should be used in unit tests, in methods it does not provide clearer logging
Suggestion: 

```
if len(labels_mask) != len(tokenized_sequence):
    raise ValueError(
        f"labels_mask length {len(labels_mask)} != tokenized_sequence length {len(tokenized_sequence)}"
    )
```

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 8]
```diff
@@ -108,12 +119,38 @@ def __len__(self) -> int:
...
+            labels_mask = None
 
         label = tokenized_sequence if self.labels is None else self.transform_label(self.labels.iloc[index])
         # Overall mask for a token being masked in some capacity - either mask token, random token, or left as-is
         loss_mask = ~torch.isin(tokenized_sequence, Tensor(self.tokenizer.all_special_ids))
 
+        # Combine with labels_mask if it exists
+        if labels_mask is not None:
+            # Convert string labels_mask to boolean tensor
+            labels_mask_tensor = torch.tensor([bool(int(char)) for char in labels_mask], dtype=torch.bool)
```
Comment (dorotat-nv) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:151]: minor performance optimisation and cleared for readability 

`labels_mask_tensor = torch.tensor([c == "1" for c in labels_mask], dtype=torch.bool)` and an even faster option for larger masks is

`labels_mask_tensor = torch.from_numpy(np.fromiter(labels_mask, dtype="U1") == "1")`

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 9]
```diff
@@ -108,12 +119,38 @@ def __len__(self) -> int:
...
+            loss_mask = loss_mask & labels_mask_tensor
```
Comment (dorotat-nv) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:152]: loss_mask &= labels_mask_tensor

Comment (dorotat-nv) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:152]: I would externalise the code to` _create_loss_mask` and unit test it if this code is not tested

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/flip_preprocess.py [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+    "meltome": ["id", "sequence", "target"],
+    "sav": ["id", "sequence", "target"],
+    "scl": ["id", "sequence", "scl_label"],
+    "secondary_structure": ["id", "sequence", "3state", "resolved"],
+}
+
+URL_PREFIX = "http://data.bioembeddings.com/public/FLIP/fasta/"
+
+
+def get_attributes_from_seq(sequences: List) -> Dict[str, Dict[str, str]]:
```
Comment (dorotat-nv) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/flip_preprocess.py:79]: could you improve the type of sequence, is it scring or any? it seems as an object with attributed name and description

is this method unittested?

File: sub-packages/bionemo-esm2/tests/bionemo/esm2/model/finetune/test_flip_preprocess.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+import tempfile
+from pathlib import Path
+
+import pytest
+
+from bionemo.esm2.model.finetune.flip_preprocess import FLIPPreprocess
+
+
+@pytest.fixture
+def temp_dir():
```
Comment (dorotat-nv) [inline sub-packages/bionemo-esm2/tests/bionemo/esm2/model/finetune/test_flip_preprocess.py:27]: lets use fixture `temp_path` available in pytest - it path is pytest run and test specific so easy to debug

File: sub-packages/bionemo-llm/src/bionemo/llm/lightning.py [hunk 12]
```diff
@@ -369,9 +369,17 @@ def update_metric(
             case "classification":
                 classification_output = outputs["classification_output"]
                 num_classes = classification_output.shape[-1]
```
Comment (dorotat-nv) [inline sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:371]: I would suggest the following changes for better clarity and and slight optimisation

```
labels = batch["labels"]
loss_mask = batch["loss_mask"].view(-1)

if classification_output.ndim == 3:  # token-level classification
    classification_output = classification_output.view(-1, num_classes)[loss_mask]
    labels = labels.view(-1)[loss_mask]

if classification_output.ndim != 2:
        raise ValueError(
            f"Expected 2D classification_output after masking, got shape {classification_output.shape}"
        )
```

Can we unit tests it as well?

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 5 above)
Comment (dorotat-nv) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:134]: please unit test _update_labels_maks_with_special_tokens

**Comment by trvachov** [comment]
Ok -- @yzhang123 I suggest to look over @dorotat-nv 's comment, however at this point we should possibly roll them up in a follow up PR, because I would like to pull these changes in to begin QA testing tomorrow (which is a workday in Shanghai, so we can get 1 day of QA testing before next week).

File: sub-packages/bionemo-llm/src/bionemo/llm/lightning.py (hunk 12 above)
Comment (trvachov) [reply sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:371]: Will do in separate PR @yzhang123

File: sub-packages/bionemo-esm2/tests/bionemo/esm2/model/finetune/test_flip_preprocess.py (hunk 11 above)
Comment (trvachov) [reply sub-packages/bionemo-esm2/tests/bionemo/esm2/model/finetune/test_flip_preprocess.py:27]: Will do in separate PR @yzhang123

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/flip_preprocess.py (hunk 10 above)
Comment (trvachov) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/flip_preprocess.py:79]: Will do in separate PR @yzhang123

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 9 above)
Comment (trvachov) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:152]: Will do in separate PR @yzhang123

Comment (trvachov) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:152]: Will do in separate PR @yzhang123

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 8 above)
Comment (trvachov) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:151]: Will do in separate PR @yzhang123

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 7 above)
Comment (trvachov) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:138]: Will do in separate PR @yzhang123

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 6 above)
Comment (trvachov) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:127]: Will do in separate PR @yzhang123

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 5 above)
Comment (dorotat-nv) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:134]: @yzhang123 , I was thinking about this one and IMHO inserting mask tokens for special tokens should be a method in the tokeniser since the mechanism of how the special tokens are inserted is tokenizer specific

Comment (yzhang123) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:134]: yes A is correct

Comment (yzhang123) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:134]: here is the test https://github.com/NVIDIA/bionemo-framework/pull/946/files#diff-2622787f13507b1b066408612713137d413f87d57fc6ac6d274996b1dc0064c5
